### PR TITLE
Correctly fold the narrowing casts

### LINKED ISSUE #7825: const uint is not trimmed
```nim
const c = (not uint32(0)) shl 8
let l = (not uint32(0)) shl 8
echo "c: ", int64(c)
echo "l: ", int64(l)
doAssert(c == l)
```
```
c: -256
l: 4294967040
test.nim(5)              test
system.nim(3781)         raiseAssert
system.nim(2843)         sysFatal
Error: unhandled exception: int64(-256) == int64(l)  [AssertionError]
```

### PR DESCRIPTION
**Pull request by LemonBoy**
Fixes #7825, #2478, #7300 and supersedes #8062.

CC @Araq

### CODE REVIEW

File: compiler/semfold.nim [hunk 1]
```diff
@@ -211,12 +211,7 @@ proc evalOp(m: TMagic, n, a, b, c: PNode; g: ModuleGraph): PNode =
   of mUnaryMinusF64: result = newFloatNodeT(- getFloat(a), n, g)
   of mNot: result = newIntNodeT(1 - getInt(a), n, g)
   of mCard: result = newIntNodeT(nimsets.cardSet(g.config, a), n, g)
-  of mBitnotI:
```
Comment (Araq) [inline compiler/semfold.nim]: Huh, reverting these bugfixes?

Comment (LemonBoy) [reply compiler/semfold.nim]: If the narrowing casts are now correctly folded we don't need to perform them by hand after every operations

**Comment by LemonBoy** [comment]
No idea about the test failures, they all look unrelated to the PR :man_shrugging:

**Comment by Araq** [comment]
They look related to me: https://ci.appveyor.com/project/Araq/nim/build/4750/tests

```
tests/errmsgs/t1154.nim errmsgs 328 ms
Error message:
Failure: reMsgsDiffer
Expected:
invalid type: 'expr' in this context: 'proc (a: varargs[expr])' for proc

Gotten:
invalid type: 'expr' in this context: 'proc (a: varargs[expr]){.locks: -1.}' for proc
```

**Comment by LemonBoy** [comment]
No matter how hard I try I cannot reproduce any of those errors locally.

File: compiler/semfold.nim [hunk 2]
```diff
@@ -431,19 +424,27 @@ proc rangeCheck(n: PNode, value: BiggestInt; g: ModuleGraph) =
...
+  let dstTyp = skipTypes(a.typ, abstractRange)
+
+  case srcTyp.kind
   of tyInt..tyInt64, tyUInt..tyUInt64:
-    case skipTypes(a.typ, abstractRange).kind
+    case dstTyp.kind
     of tyFloat..tyFloat64:
       result = newIntNodeT(int(getFloat(a)), n, g)
     of tyChar: result = newIntNodeT(getOrdValue(a), n, g)
+    of {tyUInt8..tyUInt32}:
```
[the remark this answers is not in the document]
Comment (LemonBoy) [reply compiler/semfold.nim]: Nope, the same error popped up again

**Comment by LemonBoy** [comment]
Ok, I've rewritten the patch from scratch with better understanding of wtf went wrong.

The real problem is that `winlean` is wrong as you can see from the failing test: a `DWORD` [should be](https://docs.microsoft.com/en-us/windows/desktop/winprog/windows-data-types) a `uint32` and not a `int32`. I'm no Windows expert so I'm putting this on hold until somebody verifies and patches that.

**Pull request merged**